command/debug.go

```diff
@@ -522,331 +502,286 @@ func (c *DebugCommand) captureStaticTargets(client *api.Client) error {
-			case <-c.doneCh:
-				return
-			}
-
-			// We want to add 2 at a time to ensure that we capture both at the
-			// same interval slice.
-			c.pprofWg.Add(2)
+			var wg sync.WaitGroup
+			wg.Add(2)
 
 			// Capture profile
 			go func() {
-				defer c.pprofWg.Done()
-				data, err := pprofProfile(client, c.flagInterval)
+				defer wg.Done()
+				data, err := pprofProfile(c.ctx, c.client, c.flagInterval)
 				if err != nil {
-					errCh <- newCaptureError("pprof.profile", err)
+					c.recordError("pprof.profile", err)
 					return
 				}
 
 				err = ioutil.WriteFile(filepath.Join(dirName, "profile.prof"), data, 0644)
 				if err != nil {
-					errCh <- newCaptureError("pprof.profile", err)
+					c.recordError("pprof.profile", err)
 				}
 			}()
 
 			// Capture trace
 			go func() {
-				defer c.pprofWg.Done()
-				data, err := pprofTrace(client, c.flagInterval)
+				defer wg.Done()
+				data, err := pprofTrace(c.ctx, c.client, c.flagInterval)
 				if err != nil {
-					errCh <- newCaptureError("pprof.trace", err)
+					c.recordError("pprof.trace", err)
 					return
 				}
 
 				err = ioutil.WriteFile(filepath.Join(dirName, "trace.out"), data, 0644)
 				if err != nil {
-					errCh <- newCaptureError("pprof.trace", err)
+					c.recordError("pprof.trace", err)
 				}
 			}()
 
-			// Wait until profile/trace is done
-			c.pprofWg.Wait()
-		}()
+			wg.Wait()
+		}
 	}
+}
 
-	if strutil.StrListContains(c.flagTargets, "replication-status") {
-		c.logger.Info("capturing replication status", "count", idxCount)
-
-		wg.Add(1)
-		go func() {
-			r := client.NewRequest("GET", "/v1/sys/replication/status")
-			resp, err := client.RawRequest(r)
+func (c *DebugCommand) replStatusCapture() error {
+	ticker := time.Tick(c.flagInterval)
+	for {
+		select {
+		case <-c.ctx.Done():
+			return nil
+		case <-ticker:
+			currentTimestamp := time.Now().UTC()
+			r := c.client.NewRequest("GET", "/v1/sys/replication/status")
+			resp, err := c.client.RawRequestWithContext(c.ctx, r)
 			if err != nil {
-				errCh <- newCaptureError("replication-status", err)
+				c.recordError("replication-status", err)
 			}
 			if resp != nil {
 				defer resp.Body.Close()
 
 				secret, err := api.ParseSecret(resp.Body)
 				if err != nil {
-					errCh <- newCaptureError("replication-status", err)
+					c.recordError("replication-status", err)
 				}
 				if replicationEntry := secret.Data; replicationEntry != nil {
 					replicationEntry["timestamp"] = currentTimestamp
-					c.replicationStatusCh <- replicationEntry
+					c.replicationStatus = append(c.replicationStatus, replicationEntry)
 				}
 			}
-			wg.Done()
-		}()
+		}
 	}
+}
 
-	if strutil.StrListContains(c.flagTargets, "server-status") {
-		c.logger.Info("capturing server status", "count", idxCount)
+func (c *DebugCommand) serverStatusCapture() error {
+	ticker := time.Tick(c.flagInterval)
+	for {
+		select {
+		case <-c.ctx.Done():
+			return nil
+		case <-ticker:
+			currentTimestamp := time.Now().UTC()
```

Comment:
the ticker returns a time.Time that can be used here

Reply:
Ah, good tip, I'll bear that in mind.